### PR TITLE
more misc work

### PR DESCRIPTION
**Pull request by est77**
- OSL flip handedness.
- Reverted removal of Ashikmin OSL closure.
- Small stdosl.h changes + formatting.

### CODE REVIEW

File: src/appleseed/foundation/math/matrix.h [hunk 1]
```diff
@@ -2028,6 +2031,12 @@ inline void Matrix<T, 4, 4>::decompose(
 }
 
 template <typename T>
+inline bool swaps_handedness(const Matrix<T, 4, 4>& mat)
+{
+    return det(mat.extract_matrix3()) < T(0);
```
Comment (dictoon) [inline src/appleseed/foundation/math/matrix.h]: `T(0.0)`

File: src/appleseed/renderer/kernel/shading/shadingpoint.cpp [hunk 2]
```diff
@@ -824,6 +824,21 @@ void ShadingPoint::initialize_osl_shader_globals(
...
+        if (m_assembly_instance_transform_seq->can_swap_handedness())
+        {
+            ass_inst_swaps_handedness =
+                m_assembly_instance_transform.swaps_handedness();
+        }
+
+        const bool obj_inst_swaps_handedness =
+            m_object_instance->transform_swaps_handedness();
+
+        m_shader_globals.flipHandedness =
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/shadingpoint.cpp]: Collapse into a single line.

File: src/appleseed/renderer/meta/tests/test_transformsequence.cpp [hunk 3]
```diff
@@ -869,4 +869,45 @@ TEST_SUITE(Renderer_Utility_TransformSequence)
             sequence,
             bbox);
     }
+
+    TEST_CASE(TestCanSwapsHandedness)
```
Comment (dictoon) [inline src/appleseed/renderer/meta/tests/test_transformsequence.cpp]: Split into multiple tests.

File: src/appleseed/renderer/modeling/scene/objectinstance.h [hunk 4]
```diff
@@ -187,6 +190,7 @@ class APPLESEED_DLLSYMBOL ObjectInstance
     foundation::uint32  m_vis_flags;
     RayBiasMethod       m_ray_bias_method;
     double              m_ray_bias_distance;
+    bool                m_transform_swaps_handedness;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/objectinstance.h]: Assign a default value in the constructor.

File: src/appleseed/foundation/math/matrix.h (hunk 1 above)
Comment (dictoon) [inline src/appleseed/foundation/math/matrix.h]: Fully inline `det(mat.extract_matrix3())`.

**Comment by est77** [comment]
Optimized a bit, addressed comments and rebased.

File: src/appleseed/renderer/utility/transformsequence.cpp [hunk 5]
```diff
@@ -270,6 +300,8 @@ void TransformSequence::copy_from(const TransformSequence& rhs)
             m_interpolators[i] = rhs.m_interpolators[i];
     }
     else m_interpolators = 0;
+
+    m_can_swap_handedness = rhs.m_can_swap_handedness;
```
Comment (dictoon) [inline src/appleseed/renderer/utility/transformsequence.cpp]: Missing copy of `m_all_swap_handedness`?

File: src/appleseed/renderer/utility/transformsequence.h [hunk 6]
```diff
@@ -97,6 +97,16 @@ class APPLESEED_DLLSYMBOL TransformSequence
     // Returns true on success, false otherwise.
     bool prepare();
 
+    // Return true if any of the transforms in the sequence swaps the handedness.
+    // This method can only be called after prepare has been called().
```
Comment (dictoon) [inline src/appleseed/renderer/utility/transformsequence.h]: "...after prepare() has been called."

File: src/appleseed/renderer/utility/transformsequence.h [hunk 7]
```diff
@@ -97,6 +97,16 @@ class APPLESEED_DLLSYMBOL TransformSequence
...
+    bool can_swap_handedness() const;
+
+    // Return true if the transform swaps the handedness.
+    // This method can only be called after prepare has been called()
```
Comment (dictoon) [inline src/appleseed/renderer/utility/transformsequence.h]: "...after prepare() has been called"

File: src/appleseed/renderer/utility/transformsequence.cpp [hunk 8]
```diff
@@ -198,9 +203,34 @@ bool TransformSequence::prepare()
         }
     }
 
+    m_all_swap_handedness = true;
```
Comment (dictoon) [inline src/appleseed/renderer/utility/transformsequence.cpp]: You should also reset `m_can_swap_handedness` here.

File: src/appleseed/renderer/utility/transformsequence.h [hunk 9]
```diff
@@ -97,6 +97,16 @@ class APPLESEED_DLLSYMBOL TransformSequence
     // Returns true on success, false otherwise.
     bool prepare();
 
+    // Return true if any of the transforms in the sequence swaps the handedness.
```
Comment (dictoon) [inline src/appleseed/renderer/utility/transformsequence.h]: "...if at least one of the transforms in the sequence..."

**Comment by est77** [comment]
Fixed and rebased

**Comment by dictoon** [comment]
Thanks!

**Pull request merged**